src/util/importJSON.ts

```diff
@@ -0,0 +1,209 @@
+import _ from 'lodash'
+import { Child, Context, Lexeme, ParentEntry, Path } from '../types'
+import { State } from './initialState'
+import { EM_TOKEN, ROOT_TOKEN } from '../constants'
+import { getRankAfter, getThought, getThoughts, nextSibling } from '../selectors'
+
+// util
+import {
+  addThought,
+  contextOf,
+  createId,
+  equalPath,
+  equalThoughtRanked,
+  hashContext,
+  hashThought,
+  head,
+  headRank,
+  pathToContext,
+  removeContext,
+  rootedContextOf,
+  timestamp,
+} from '../util'
+import { Block } from '../action-creators/importText'
+import { GenericObject } from '../utilTypes'
+
+interface ImportHtmlOptions {
+  skipRoot? : boolean,
+}
+
+interface RankInfo {
+  rank: number,
+  deepLevel: number,
+}
+
+/** If skipRoot is true, replace head block with its children, or drop it, if head has no children. */
+const skipRootThought = (skipRoot: boolean, blocks: Block[]) => {
+  if (!skipRoot) return blocks
+  const head = _.head(blocks)
+  if (!head) return blocks
+  const tail = _.tail(blocks)
+  return head.children.length > 0 ? [...head.children, ...tail] : tail
+}
+
+/** Calculate last thought of the first level, as this is where the selection will be restored to. */
+const calculateLastThoughtFirstLevel = (rankMap: Map<Block, RankInfo>, blocks: Block[]) => {
+  const lastThoughtFirstLevelIndex = blocks.length - 1
+  const lastThoughtFirstLevel = blocks[lastThoughtFirstLevelIndex]
+  const { rank } = rankMap.get(lastThoughtFirstLevel)!
+  return { value: lastThoughtFirstLevel.scope, rank }
+}
+
+/** Return map of thought ranks. */
+const createRankMap = (blocks: Block[], rankStart: number, rankIncrement: number) => {
```

Comment:
No, this is not the right way to do this. You have separated the ranking algorithm from the insertion algorithm and created a lot of additional computation in the process. Thoughts are already traversed in order during the insertion process, and the rank increases at regular intervals. This should not be a separate/pre-process. (Not to mention we already do a separate full traversal for `getRankIncrement`).

Does this make sense? Can you describe a better solution? I want to make sure you understand before proceeding.

Reply:
Okay, i can we can merge `createRankMap` and `saveThoughts` together. Iterate through tree once, save ranks information in `Map`, and then retrieve info in next iterations. Result function will be a bit complex and large, but we will iterate only once.

Reply:
No, we don't want to iterate a separate time, and we don't want a separate `Map`. That is effectively what you have already done, it would just be wrapped in a single function. It doesn't change the performance characteristics. You need to do it all in one iteration. Additional iterations are not necessary. Can the ranks be generated within `insertThought`? That's what we want.

Reply:
Actually, i don't think that it's a good idea. Purpose of `insertThought` is saving thoughts, and i don't think that it needs additional computation inside it. I like that it takes minimum required arguments: `value`, `context` and `rank` and saves thought. It's simple function and its purpose is clear. 
Maybe you don't understand me correctly, but `saveThoughts` and `calculateRanks` are the same in terms of iteration. They both recursively iterate through tree and do some operations on each iteration. I meant that we can substitute these 2 functions with single one, because they merge easily. On first level of recursion we can create `Map` for storing previous ranks and use this `Map` only inside this single function.
We need to know the rank of previous saved thought and how deep this thought is inside the tree.
E.g. If we recursively iterate such tree
- a
  - b
    - c
- d

on `d` thought iteration we need to know `RankInfo` of `c` thought. It could be retrieved from global mutating object (this approach was used in previous version, where `let` variable was used for it), or Map.

Reply:
> Actually, i don't think that it's a good idea. Purpose of `insertThought` is saving thoughts, and i don't think that it needs additional computation inside it. I like that it takes minimum required arguments: `value`, `context` and `rank` and saves thought. It's simple function and its purpose is clear.

Ah, yes, you're right, `insertThought` is not the place for it.

> Maybe you don't understand me correctly, but `saveThoughts` and `calculateRanks` are the same in terms of iteration. They both recursively iterate through tree and do some operations on each iteration. I meant that we can substitute these 2 functions with single one, because they merge easily.

Yes! Exactly.

> We need to know the rank of previous saved thought and how deep this thought is inside the tree.

Ranks are only relative to siblings, as I have documented and we have discussed in depth elsewhere. So they don't need to know how deep the thought is inside the tree. They just need to know the index within their parent. The previous implementation autoincremented the whole tree for efficiency. However, it's not the only possible implementation that meets the rank requirements. If it's more efficient now to increment siblings only, do that. Or if there is an efficient way to do a global autoincrement, that would be fine, too. Either way, the rank should be generated within the main traversal function, as you pointed out.

> On first level of recursion we can create `Map` for storing previous ranks and use this `Map` only inside this single function. We will iterate tree only **once**.

(I quoted this out of order since it may be invalidated by the above paragraph. Feel free to only respond to these questions if they are still relevant.)

By "first level of recursion" do you mean the first set of children? i.e. just "when the iteration starts"? I thought you meant "first iteration", my bad.

If you're only iterating the tree once, what is the purpose of `Map`? If the ranks are generated on the fly, I don't see why you would need to store them in a global structure. The starting rank can be threaded through the traversal and incremented as needed, unless I'm mis-assessing the situation.

Reply:
Okay, i was a bit confused with previous ranking implementation and that fact that some tests is based on this implementation. If it's not necessary to increment rank each time we save thought/pop out of context, then i would like to implement ranking based on siblings
Instead of
- a (0)
  - b (1)
    - c (2)
- d (5)
- d (6)

it will be (of course ranking could be a little different in "real life", `startRank` and `rankIncrement` could differ from this example, but i hope you get the idea)
- a (0)
  - b (0)
    - c (0)
- d (1)
- e (2)

Such implementation is much more easier and doesn't require any additional variables for storing previous ranks.

Reply:
💯

Reply:
> some tests is based on this implementation.

Yes, this will be a good opportunity to remove those misleading dependencies. Tests that do not rely on specific ranks should generally not be testing specific ranks. We can use `matchObject` instead of `toEqual` in some cases. I also created [`setCursorFirstMatch`](https://github.com/cybersemics/em/blob/dev/src/test-helpers/setCursorFirstMatch.ts) so that we could `setCursor` without depending on rank.